plotly/offline.py

```diff
@@ -0,0 +1,178 @@
+""" Plotly Offline
+    A module to use Plotly's graphing library with Python
+    without connecting to a public or private plotly enterprise
+    server.
+"""
+from __future__ import absolute_import
+
+import uuid
+import json
+import os
+import requests
+
+import plotly.plotly as py
+from plotly import utils
+from plotly.exceptions import PlotlyError
+
+PLOTLY_OFFLINE_DIRECTORY = plotlyjs_path = os.path.expanduser(
+    os.path.join(*'~/.plotly/plotlyjs'.split('/')))
+PLOTLY_OFFLINE_BUNDLE = os.path.join(PLOTLY_OFFLINE_DIRECTORY,
+                                     'plotly-ipython-offline-bundle.js')
+
+__PLOTLY_OFFLINE_INITIALIZED = False
+
+
+def download_plotlyjs(download_url):
+    if not os.path.exists(plotlyjs_path):
+        os.makedirs(plotlyjs_path)
+
+    res = requests.get(download_url)
+    res.raise_for_status()
+
+    with open(PLOTLY_OFFLINE_BUNDLE, 'wb') as f:
+        f.write(res.content)
+
+    print('\n'.join([
+        'Success! Now start an IPython notebook and run the following ' +
+        'code to make your first offline graph:',
+        '',
+        'import plotly',
+        'plotly.offline.init_notebook_mode() '
+        '# run at the start of every ipython notebook',
+        'plotly.offline.iplot([{"x": [1, 2, 3], "y": [3, 1, 6]}])'
+    ]))
+
+
+def init_notebook_mode():
```

Comment:
yeah, here's a scenario where embedding `init_notebook_mode` would cause unexpected things:

1 - User calls iplot. we check module level variable, see that plotly.js hasn't been inserted, so we drop it in
2 - User is messing around, so deletes the cell. plotly.js is already in memory, so no big deal, iplot will still work. User calls iplot again, since we've already imported plotly.js, we don't insert again.

3 - User shares or re-opens the notebook. The original output cell that contained plotly.js was deleted from (2), so now Plotly is undefined

similarly if someone calls iplot, embedding plotly.js, and then makes a plot in the cell *above* the first cell - when re-opening the notebook the first cell containing the `Plotly.plot` call will be before plotly.js was embedded. (This could still occur with `init_notebook_mode` - the user could call `init_notebook_mode` on cell 10, then call iplot later on cell 3, but this way is a little more explicit)

I also thought about just calling this function when we `import` plotly.offline (since that'll necessarily be the first thing called). however, we don't want to attempt to import the JS before someone calls `download_plotlyjs`...

it's definitely annoying though, I'm not sure if there is a better way to do it.